### ISSUE TITLE
[Bug]: H100 FA3 violates batch invariance when an unrelated prefill changes tiling

### ISSUE THREAD

**Issue by Yunzez** [vllm-project/vllm#51902]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
GPU: NVIDIA H100 80GB HBM3, compute capability 9.0
Driver: 595.71.05
vLLM: 0.25.1
PyTorch: 2.11.0+cu130
CUDA reported by PyTorch: 13.0
cuDNN: 9.19.0
dtype: bfloat16
attention shape: 14 Q heads, 2 KV heads, head dim 64
paged KV block size: 16 (primary), 128 (boundary control)
KV sequence length: 19802
```

</details>

### 🐛 Describe the bug

With vLLM 0.25.1 on H100, `VLLM_BATCH_INVARIANT=1` does not make the FlashAttention 3 path batch invariant for mixed decode+prefill batches. In a legal two-sequence varlen call, the same paged-KV decode row produces different BF16 output when an unrelated prefill row grows from 9 to 10 tokens. The victim's Q, KV, sequence length, and physical block mapping do not change; the neighbor's first nine Q/K/V tokens are also identical.

The scalar changes FA3's SM90 specialization:

- `max_seqlen_q=9`:   `VarlenDynamicPersistentTileScheduler<64,192,...>`
- `max_seqlen_q=10`:   `VarlenDynamicPersistentTileScheduler<192,128,...>`

The calls use exact, self-consistent maxima:
`cu_seqlens_q=[0,1,10], max_seqlen_q=9` and `cu_seqlens_q=[0,1,11], max_seqlen_q=10`. 
They both use `num_splits=1` and disjoint physical KV blocks. Repeated calls with identical complete inputs are bitwise stable, so this is a deterministic cross-tiling numerical difference, not stochastic behavior within one kernel.

The arithmetic mechanism is localized to the changed N-block partition and online-softmax/P-times-V accumulation tree. With the same seeded tensors, the outputs are identical through KV length 128 and first differ at length 129,
while the FP32 LSE remains bitwise equal. At that boundary the N=128 specialization performs two K-block updates (128+1), while the N=192 specialization performs one. For longer inputs the LSE can differ as well.

## Why this is an issue: 

This violates the advertised contract
`FlashAttentionBackend.supports_batch_invariance()` returns `True`. In BI mode, vLLM disables FA3 AOT scheduling and forces `max_num_splits=1`, but the forward path still passes the batch-global `attn_metadata.max_query_len` unchanged as `max_seqlen_q` to `_vllm_fa3_C.fwd`.

Consequently, a decode request with unchanged Q/K/V and KV block mapping can change its attention result when a co-batched prefill raises the global maximum query length.

## Minimal reproduction

Self-contained; runs on a single H100 with vLLM 0.25.1 installed. It calls `flash_attn_varlen_func` directly and never goes through vLLM's serving path or its batch-invariance machinery, so no environment variables are needed.

```python
import math
import torch
from vllm.vllm_flash_attn.flash_attn_interface import flash_attn_varlen_func

DEV, DT = "cuda", torch.bfloat16
NQ, NKV, HD, BLOCK, VICTIM_KLEN = 14, 2, 64, 16, 19_802   # GQA ratio 7, head dim 64

def randn(seed, shape):
    g = torch.Generator(device=DEV).manual_seed(seed)
    return torch.randn(shape, generator=g, device=DEV, dtype=DT)

victim_blocks, neighbor_blocks = (VICTIM_KLEN + BLOCK - 1) // BLOCK, 1
victim_q       = randn(31, (1, NQ, HD))
victim_cache   = randn(32, (victim_blocks, 2, BLOCK, NKV, HD))
neighbor_q     = randn(33, (10, NQ, HD))
neighbor_cache = randn(34, (neighbor_blocks, 2, BLOCK, NKV, HD))
key_cache, value_cache = torch.cat((victim_cache, neighbor_cache), 0).unbind(1)

victim_ids   = torch.arange(victim_blocks, device=DEV, dtype=torch.int32)
neighbor_ids = torch.arange(victim_blocks, victim_blocks + neighbor_blocks,
                            device=DEV, dtype=torch.int32)
assert not torch.isin(victim_ids, neighbor_ids).any()   # disjoint physical blocks

def run(neighbor_len):
    """Victim is row 0: a 1-token decode whose every input is fixed."""
    q      = torch.cat((victim_q, neighbor_q[:neighbor_len]), 0)
    cu_q   = torch.tensor([0, 1, 1 + neighbor_len], device=DEV, dtype=torch.int32)
    used_k = torch.tensor([VICTIM_KLEN, neighbor_len], device=DEV, dtype=torch.int32)
    table  = torch.full((2, victim_blocks), -1, device=DEV, dtype=torch.int32)
    table[0, :victim_blocks]   = victim_ids
    table[1, :neighbor_blocks] = neighbor_ids

    # max_seqlen_q is exactly the longest row present; nothing is over-declared.
    assert neighbor_len == int((cu_q[1:] - cu_q[:-1]).max())

    out = flash_attn_varlen_func(
        q, key_cache, value_cache,
        cu_seqlens_q=cu_q, seqused_k=used_k,
        max_seqlen_q=neighbor_len, max_seqlen_k=VICTIM_KLEN,
        block_table=table, softmax_scale=1 / math.sqrt(HD),
        causal=True, fa_version=3, num_splits=1,
    )
    torch.cuda.synchronize()
    return out[0].clone()          # victim row only

for _ in range(5):
    run(9); run(10)

a, b = run(9), run(10)
print("victim bitwise equal :", torch.equal(a, b))
print("unequal BF16 elements:", int((a != b).sum()))
print("max abs delta        :", float((a.float() - b.float()).abs().max()))
```

Only the neighbor's length differs between the two calls, and the neighbor's first nine Q/K/V tokens are shared by both.

Observed victim result for a fixed seeded input:

```text
legal neighbor q=9 vs q=10: 95 unequal BF16 elements
maximum abs delta:           0.00006103515625
same-input stability:        30/30 bitwise-identical runs for each case
solo max_q=9 == legal q=9:   bitwise equal
solo max_q=10 == legal q=10: bitwise equal
```

The neighbor's first nine Q/K/V tokens are shared between both cases, unused block-table entries are `-1`, and victim/neighbor physical blocks are asserted disjoint. The earlier one-row scalar probe remains useful as selector localization, but is not needed for the BI claim.

The kernel identities above come from `torch.profiler` tables rendered with `max_name_column_width` raised past the 55-character default, so the full demangled scheduler template parameters are visible rather than truncated.
Untruncated tables were captured for the single-row, legal-two-row, and head-dim-128 cases; they can be attached on request.

## Numerical-stage evidence

Additional controls use the same one-row paged-KV call:

- random Q/K/V: 108 unequal BF16 outputs, but FP32 LSE is bitwise equal;
- zero Q (uniform logits): output and LSE are bitwise equal;
- constant V: output and LSE are bitwise equal;
- 64 single-position V impulses across both tile boundaries: output and LSE
  are bitwise equal;
- both tilings have almost identical error relative to an FP64 reference.

The K=129 boundary is the concise causal localization: the two separately compiled tile shapes induce different blockwise reduction topologies. No single shared "bad instruction" is required to explain the result.

## Boundary generalization

A legal-input sweep tested the selector formula `max_seqlen_q * (nq / nkv) <= 64` over five seeds per case:

| GQA ratio | tested boundary | seeds with victim difference |
| ---: | ---: | ---: |
| 1 | 64 -> 65 | 5/5 |
| 2 | 32 -> 33 | 5/5 |
| 4 | 16 -> 17 | 5/5 |
| 7 | 9 -> 10 | 5/5 |
| 8 | 8 -> 9 | 5/5 |

All exact-input repeats were bitwise stable. This establishes agreement with the source formula in the tested configurations; it is not a claim that every GQA model or input necessarily produces an observable numerical difference.

With KV page size 128, the primary ratio-7 9 -> 10 violation also occurred in 5/5 seeds. A second source-predicted PagedKV TMA dispatch transition at 27 -> 28 changed one compiled specialization boolean but produced no victim numerical difference in 5/5 seeds. Other page sizes can therefore expose additional dispatch boundaries, not necessarily additional BI violations.

### Head-dim scope control

The same legal two-row sweep was repeated with head dim 128 for GQA ratios 1, 2, 4, 7, and 8, using five seeds and three repeats per exact input. Every boundary was 0/5 numerically different. Profiling showed the expected transitions:

```text
use_one_mma_wg=true:  (kBlockM,kBlockN) = (64,128)
use_one_mma_wg=false: (kBlockM,kBlockN) = (128,128)
```

Thus only the Q-side M tile changes; the KV N partition remains identical. The page-128 ratio-7 9 -> 10 control was also 0/5 different. This report therefore limits the demonstrated numerical violation to the tested head-dim-64 path and
does not advertise it as a general head-dim-128 model bug. The negative control is evidence for the current N-block/PV-association mechanism, not a proof that every possible d=128 input is mathematically identical. For d=128, page size 128 has a separate PagedKV TMA boundary at q=18 -> 19, because `max_q * 7` crosses `kBlockM=128` there; q=27 -> 28 is already on the same side. That
18 -> 19 transition was tested over five seeds and three repeats per exact input: the predicted TMA boolean changed, while the victim remained bitwise equal in 5/5 seeds.

The selector still causes a performance regression for d=128 even when the outputs are equal. At K=19,802, the measured `64x128` versus `128x128` median was 176.23 -> 244.98 us (1.390x); at K=2,048 it was 25.26 -> 33.10 us (1.310x),
and at K=8,192 it was 77.43 -> 106.29 us (1.373x). Therefore the correctness scope demonstrated here is d=64, while the decode-performance scope includes both d=64 and d=128.

Concrete d=64 examples include Qwen2.5-0.5B, Llama-3.2-1B, and TinyLlama-1.1B.
The model list is illustrative; the applicable condition is the model's attention head dimension, not its name or parameter count.

Speculative decoding is a particularly plausible high-frequency trigger. A verification step commonly has `q_len = proposed_tokens + 1`; for ratio 7, `q_len >= 10` crosses the tested selector boundary, and even shorter verify requests can be affected when co-batched with a prefill whose global maximum crosses it. This is a mechanism-based scope implication, not a separate speculative-decoding reproduction.

<details>
<summary> Click here to see the evidence</summary>

## Some service-level evidence

- On vLLM 0.25.1, the exact concurrent Trace reproduced under FA3+BI with verified cache reset in 4/4 comparisons; the same Trace under TRITON_ATTN+BI was stable in 0/4 comparisons. Both new evidence files retain the exact Trace/payload hashes and full run provenance.
- A one-layer eager hook counted 527 real FA3 engine-step calls during one exact Trace: 510 had `max_seqlen_q <= 9`, while 17 had `max_seqlen_q >= 10` (3.23%, maximum 8192). The workload therefore crosses the selector gate repeatedly; this histogram is explicitly eager because graph replay bypasses Python hooks.
- A vLLM 0.25.1 graph-mode replay with r1 logprobs produced three distinct r1 outputs in five runs. All four comparisons first changed at output-token index 18. To be precise about the strength of this particular flip: the reference run's top1-top2 gap at that position was 0, i.e. an exact argmax tie, so on the taxonomy used in #47069 this is a tie flip rather than a confident flip. (The 0.125 figures are the gaps in the *diverged* trajectories, downstream of the flip, and are not evidence that the flip  itself was confident.) A confident flip, reference gap 0.125, not a tie, was demonstrated for this same bug on vLLM 0.19.0 in #47069. The kernel mechanism is identical across both versions; what differs is which token position the workload happened to land on.
- A live FA3 hook recomputed a real batched target row alone. Using the target's local maximum differed from the batched output in 153/300 calls. Retaining  the original batch-global `max_seqlen_q/max_seqlen_k` reproduced the batched output in 300/300 calls.
- Forcing vLLM's persistent SM80 matmul overrides on SM90 did not eliminate the  finding. Independently of graph mode, an operator-level FA2 scan over `max_seqlen_q=1..17,32,...,8191` was bitwise equal throughout, including all 136 pairwise comparisons among values 10..8191; FA2 used one `flash_fwd_splitkv_kernel<...64,64,256...>` across 9/10.
- As a causal ablation, clamping BI+FA3 `max_seqlen_q` to at least 10 kept both decode-only and mixed calls on the `192x128` side. The exact finding then became stable for all 11 deterministic requests across 20 runs (19/19 comparisons per request), with verified HTTP-200 cache resets before every run. This remains a causal ablation, not a proposed production fix. The  slower forced specialization can shift later batch timing, so timeline displacement remains a residual service-level confound; the standalone legal-input probes supply the scheduling-free causal evidence.
- The d=128 performance benchmark used nine alternating samples and 500  iterations per sample. It found a 1.390x slowdown at K=19,802 when the batch-global selector moved decode from `64x128` to `128x128`, despite bitwise-equal outputs. This makes the performance impact broader than the demonstrated d=64 correctness impact.
- A real service profile contained 52,558 `matmul_kernel_persistent*` launches and no cuBLAS/cuBLASLt/GEMM kernel-name matches. Together with the installed source path, this shows that an independent direct native-`F.linear` Hopper observation is not reachable through the tested unquantized Qwen2 BI service and is not part of this report. In the standalone control, each fixed M was bitwise stable over 30 repeats; cross-M differences reproduced under `CUBLAS_WORKSPACE_CONFIG=:16:8` (6 elements, max abs 0.125) but disappeared under `:4096:8`, further supporting its treatment as a configuration-specific latent operator note.

</details>

## Expected behavior

When batch invariance is enabled, adding an unrelated prefill row must not change the attention output of an otherwise identical decode row.

Setting the batch-global `max_seqlen_q` to 1 is not a valid fix because the same call contains actual prefill rows longer than 1. Keeping `num_splits=1` alone is already shown to be insufficient. Globally forcing the `192x128` path is also a poor production fix: at KV length 19,802 it measured 251.97 us versus 143.53 us for `64x192` on a single decode row (1.756x).

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by andakai** [comment]
https://github.com/vllm-project/flash-attention/pull/169 is trying to fix this issue. https://github.com/vllm-project/vllm/pull/48694 is also related.
